RAG Debate.py

```diff
@@ -57,48 +74,148 @@ def generate_response(query, vectorizer, vectors, paragraphs, api_key):
+        paragraph_length = estimate_tokens(paragraph, model_name)
+        if current_length + paragraph_length > max_tokens:
+            chunked_paragraphs.append((current_chunk, current_filenames))
+            current_chunk = [paragraph]
+            current_length = paragraph_length
+            current_filenames = [filename]
+        else:
+            current_chunk.append(paragraph)
+            current_length += paragraph_length
+            current_filenames.append(filename)
     
-    vectorizer, vectors = vectorize_paragraphs(processed_paragraphs)
+    if current_chunk:
+        chunked_paragraphs.append((current_chunk, current_filenames))
     
-    # Replace with OpenAI API key
-    api_key = 'INSERT API'
-
-    def chatbot():
-        while True:
-            query = input("You: ")
-            if query.lower() in ['exit', 'quit']:
-                break
-            response, retrieved_text = generate_response(query, vectorizer, vectors, processed_paragraphs, api_key)
+    return chunked_paragraphs
+
+def chatbot(processed_paragraphs, all_filenames, api_key):
+    best_response = None
+    best_similarity = 0
+    best_retrieved_text = None
+
+    chunked_paragraphs = chunk_paragraphs(processed_paragraphs, all_filenames, max_tokens=2000, model_name='gpt-4')
+    print(f"Number of chunks: {len(chunked_paragraphs)}")  # Debug statement
+    for idx, (chunk, filenames) in enumerate(chunked_paragraphs):
+        print(f"Chunk {idx + 1}: Contains {len(chunk)} paragraphs from files: {set(filenames)}")  # Debug each chunk
+    
+    while True:
+        query = input("You: ")
+        if query.lower() in ['exit', 'quit']:
+            break
+
+        all_responses = []
+        found_relevant_info = False
+        best_response = None
+        best_similarity = 0
+        best_retrieved_text = None
+        
+        for i, (chunk, filenames) in enumerate(chunked_paragraphs):
+            if not chunk:
+                continue
+            vectorizer, vectors = vectorize_paragraphs(chunk)
+            if vectors.shape[0] == 0:
+                continue
+            try:
+                print(f"Processing chunk {i+1}/{len(chunked_paragraphs)} from files: {set(filenames)}")  # Debug statement
+                query_vec = vectorizer.transform([query])
+                similarities = cosine_similarity(query_vec, vectors).flatten()
+                idx = np.argmax(similarities)
+                if similarities[idx] > best_similarity:
+                    best_similarity = similarities[idx]
+                    best_retrieved_text = chunk[idx]
+                    best_response = generate_response(query, vectorizer, vectors, chunk, api_key, max_tokens=2000)
+                    found_relevant_info = True
+            except ValueError as e:
+                print(e)
+                continue
+
+        if not found_relevant_info:
+            print("No relevant info found, reprocessing documents...")  # Debug statement

```

Comment:
why is it useful to reprocess documents when it just looked through all of them and found them to be all irrelevant? Isn't it better to just generate an "i don't have enough information to answer this" response?